File: 01-programming/csharp/concepts/code/01-basics/datatype-floating-points.cs
// -o-o-o-o- Datatype (Floating Points) -o-o-o-o-

/* This file goes over floating point data types in C# simply.
 *
 *  # # # # KEY # # # #
 *
 * float   - a 32-bit decimal number. Requires an 'f' suffix on literals.
 *           eg. float x = 3.14f;
 *           Precise to ~7 digits. Equivalent to float in C.
 *
 * double  - a 64-bit decimal number. This is the DEFAULT decimal type in C#.
 *           eg. double x = 3.14;
 *           Precise to ~15-16 digits. Equivalent to double in C.
 *
 * decimal - a 128-bit decimal number. Requires an 'm' suffix on literals.
 *           eg. decimal x = 3.14m;
 *           Precise to ~28-29 digits. Best used for money/finance.
 *           No direct equivalent in C.
 *
 * Console.WriteLine() - prints text to the console with a newline at the end.
 *
 * {0}     - a format placeholder. Works like %f in C's printf.
 *           eg. Console.WriteLine("Value: {0}", myVar);
 *           You can also use string interpolation: $"Value: {myVar}"
 *
 * */

using System;

class Program {
    static void Main() {

        float   myFloat   = 3.14f;       // 32-bit decimal, note the 'f' suffix
        double  myDouble  = 3.14159265;  // 64-bit decimal, default decimal type
        decimal myDecimal = 3.14m;       // 128-bit decimal, note the 'm' suffix

        Console.WriteLine("Float value:   {0}", myFloat);
        Console.WriteLine("Double value:  {0}", myDouble);
        Console.WriteLine("Decimal value: {0}", myDecimal);

        // Showing precision differences
        float   precisionFloat   = 1.0f / 3.0f;
        double  precisionDouble  = 1.0 / 3.0;
        decimal precisionDecimal = 1.0m / 3.0m;

        Console.WriteLine("\nPrecision comparison for 1/3:");
        Console.WriteLine("Float:   {0}", precisionFloat);
        Console.WriteLine("Double:  {0}", precisionDouble);
        Console.WriteLine("Decimal: {0}", precisionDecimal);

    }
}

File: 01-programming/python/concepts/code/01-basics/datatype-floating-points.py
# -o-o-o-o- Datatype (Floating Points) -o-o-o-o-

# This file goes over floating point data types in Python simply.
#
#  # # # # KEY # # # #
#
# float    - Python's only built-in floating point type.
#            It is a 64-bit double-precision decimal number.
#            Equivalent to double in C and C#. There is no separate float/double
#            distinction in Python — float covers both roles.
#            eg. x = 3.14
#
# complex  - Python's built-in complex number type (real + imaginary).
#            Written with a 'j' suffix for the imaginary part.
#            eg. z = 2 + 3j
#            No equivalent in standard C or C#.
#
# Decimal  - a high-precision decimal type from Python's decimal module.
#            Similar to decimal in C# or using fixed-point arithmetic.
#            Best used for financial/monetary calculations where float
#            rounding errors are not acceptable.
#            eg. from decimal import Decimal; x = Decimal("3.14")
#            Note: always pass the value as a STRING to Decimal() to avoid
#            floating point errors being baked in before Decimal sees the value.
#
# round()  - built-in function to round a float to a given number of decimals.
#            eg. round(3.14159, 2) -> 3.14
#
# print()  - prints text to the console with a newline at the end.
# f"..."   - f-string. Embeds variables directly in a string.
# :.Nf     - format specifier inside an f-string to show N decimal places.
#            eg. f"{value:.4f}" shows 4 decimal places.
#

from decimal import Decimal

# Basic float (Python's default decimal type — 64-bit double precision)
my_float = 3.14
print(f"Float value:   {my_float}")
print(f"Type:          {type(my_float)}")

# Python floats are always 64-bit (no need for a separate double type)
big_float = 3.14159265358979
print(f"\nHigh precision float: {big_float}")

# Floating point precision quirk (same in all languages)
print(f"\n0.1 + 0.2 = {0.1 + 0.2}")           # not exactly 0.3 due to binary representation
print(f"Rounded:   {round(0.1 + 0.2, 2)}")    # use round() to fix display

# Complex numbers (unique to Python)
my_complex = 2 + 3j
print(f"\nComplex number: {my_complex}")
print(f"Real part:      {my_complex.real}")
print(f"Imaginary part: {my_complex.imag}")

# Decimal for precise arithmetic (eg. money)
price   = Decimal("19.99")  # always pass as a string
tax     = Decimal("0.15")
total   = price * (1 + tax)
print(f"\nPrice:  {price}")
print(f"Tax:    {tax * 100}%")
print(f"Total:  {total}")

# Precision comparison for 1/3
float_third   = 1 / 3
decimal_third = Decimal("1") / Decimal("3")
print(f"\n1/3 as float:   {float_third}")
print(f"1/3 as Decimal: {decimal_third}")

# Formatting floats in output
value = 3.14159265
print(f"\n2 decimal places: {value:.2f}")
print(f"4 decimal places: {value:.4f}")
print(f"Scientific:       {value:.2e}")
